Adding the new resource type PrivateLinkForAzureAD

KranthiKiranerusu opened the pull request:
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [x] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

KranthiKiranerusu committed:
Adding the new resource type PrivateLinkForAzureAD

AutorestCI commented:
You don't have permission to trigger SDK Automation.
Please add yourself to ***Azure*** group from opensource portal if you are MSFT employee,
or please ask reviewer to add comment *** /openapibot sdkautomation ***.
Please ask tih@microsoft.com (or NullMDR in github) for additional help.

azuresdkci commented:
Can one of the admins verify this patch?

KranthiKiranerusu committed:
Added the example

KranthiKiranerusu committed:
Updated the example

leni-msft commented:
/azp run automation - sdk

KranthiKiranerusu committed:
Updated the definition for privateLinkforAzureAD

KranthiKiranerusu committed:
Added privatelinkresource

KranthiKiranerusu committed:
Added privatelinkresource

KranthiKiranerusu committed:
Updated definitions

AutorestCI commented:
/azp run automation - sdk

KranthiKiranerusu committed:
Updated definitions

majastrz commented on specification/azureactivedirectory/resource-manager/Microsoft.Aadiam/stable/2017-04-01/azureactivedirectory.json:
```diff
@@ -37,6 +37,117 @@
   },
   "paths": {
+    "/providers/microsoft.aadiam/privateLinkForAzureAD": {
+      "get": {
```
>get [](start = 7, length = 3)

How many private links can there be? Do you need paging here?

arluca replied:
Standard ARM limits, 800 per subscription. 

This is the same as what "private link service" resource does. 

https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits#private-link-limits

majastrz replied:
You should probably add paging via the x-ms-pageable extension. It's ok to add it even if your service only returns one page right now, but it makes adding paging to the service in the future easier.

---
In reply to: [403390060](https://github.com/Azure/azure-rest-api-specs/pull/8854#discussion_r403390060) [](ancestors = 403390060)

KranthiKiranerusu replied:
makes sense I added x-ms-pageable extension. also @majastrz , @arluca , I had to restructure the changes into separate files to keep it clean.

majastrz commented on specification/azureactivedirectory/resource-manager/Microsoft.Aadiam/stable/2017-04-01/azureactivedirectory.json:
```diff
@@ -37,6 +37,117 @@
     }
   },
   "paths": {
+    "/providers/microsoft.aadiam/privateLinkForAzureAD": {
```
>privateLinkForAzureAD [](start = 33, length = 21)

Are you planning to support non-AzureAD private links? If the answer is no, you already have AAD in the namespace, so this is kind of redundant. On top of that, other RPs are calling this `privateLinks`, so it would be great if private link setup was at least somewhat consistent.

arluca replied:
@KranthiKiranerusu: This feedback makes sense. If we want to just call it private link, that's fine. @majastrz : Would we still able to give it a "display" name of "Private Link for Azure AD"?

majastrz replied:
Yeah, should be able to. I assume you're talking about the asset display name in the portal that maps to a resource type - that is completely controlled by the UX extension that publishes the asset as far as I know. 

---
In reply to: [403360887](https://github.com/Azure/azure-rest-api-specs/pull/8854#discussion_r403360887) [](ancestors = 403360887)

arluca replied:
Hmm... So I agree with your suggestion, that said, I just realized that our "privateLinkforAzureAD" name has already been given to NRP for onboarding us to private link and they manage to get us to canary while under lockdown. If we have to change our name now, it will probably delay is a few more weeks.

Is the problem with the name purely about the redundant wording?

majastrz replied:
This is why RP teams should be getting swaggers approved (or at least start the reviews) before any work is done. This way feedback can actually be incorporated.

The good news is that the situation is better than I thought initially. I looked over how other RPs are doing this and there is standardization on /privateLinkResources (which you're doing also), but the parent resource type is always different. I don't particularly like the name, but I think it's less of an issue at this point. The other rule you're violating with `privateLinkForAzureAd` that the name is singular - they're supposed to be plural (like `subscriptions`, `resourceGroups`, or `virtualMachines`), but I guess it'll have to stay like this now...

---
In reply to: [403389682](https://github.com/Azure/azure-rest-api-specs/pull/8854#discussion_r403389682) [](ancestors = 403389682)

arluca replied:
RE: "This is why RP teams should be getting swaggers approved (or at least start the reviews) before any work is done. This way feedback can actually be incorporated."

Totally valid. That was our fault for not checking in earlier. JEDI + COVID made our wires get crossed on order for this. 

In the future, would we be able to rename it? Might be a bigger problem at that point?

majastrz replied:
Since it's a tenant-level proxy resource, it would be fairly easy to make one underlying resource appear under two different names in ARM. How NRP would deal with this, I have no idea. 

---
In reply to: [403395988](https://github.com/Azure/azure-rest-api-specs/pull/8854#discussion_r403395988) [](ancestors = 403395988)

KranthiKiranerusu replied:
@majastrz , this is totally new to us and we are pressed against deadlines from JEDI and NRP alike. since their onboarding is in progress we woudl like to go ahead with privateLinkForAzureAD.  If an opprotunity presents in the next 4 weeks we will add a new resource with your suggestion and deprecate the old one while in private preivew.

KranthiKiranerusu committed:
Spell check

AutorestCI commented:
/azp run automation - sdk

majastrz commented on specification/azureactivedirectory/resource-manager/Microsoft.Aadiam/stable/2017-04-01/examples/AzureADPrivateLinkResourceListGet.json:
```diff
@@ -0,0 +1,29 @@
+      "body": {
+        "value": [
+          {
+            "id": "subscriptions/00000000-1111-2222-3333-444444444444/resourceGroups/Default/providers/Microsoft.DocumentDb/databaseAccounts/ddb1/privateLinkResources/sql",
```
>Microsoft.DocumentDb/databaseAccounts [](start = 103, length = 37)

`type` doesn't match `id` in this example. Copy/paste issue?

KranthiKiranerusu replied:
Fixed it.

majastrz commented on specification/azureactivedirectory/resource-manager/Microsoft.Aadiam/stable/2017-04-01/examples/AzureADPrivateLinkResourceGet.json:
```diff
@@ -0,0 +1,24 @@
+  "responses": {
+    "200": {
+      "body": {
+        "id": "subscriptions/00000000-1111-2222-3333-444444444444/resourceGroups/Default/providers/Microsoft.DocumentDb/databaseAccounts/ddb1/privateLinkResources/sql",
```
>Microsoft.DocumentDb/databaseAccounts [](start = 99, length = 37)

`type` doesn't match `id`

KranthiKiranerusu replied:
fixed it.

majastrz commented on specification/azureactivedirectory/resource-manager/Microsoft.Aadiam/stable/2017-04-01/azureactivedirectory.json:
```diff
@@ -37,6 +37,117 @@
+        "description": "Gets the private link resources that need to be created for a policy of AzureAD.",
+        "operationId": "PrivateLinkResources_ListByPrivateLinkPolicy",
+        "x-ms-pageable": {
+          "nextLinkName": null
```
>null [](start = 26, length = 4)

Can we add paging here too? (It's ok even if the service returns one page currently.)

KranthiKiranerusu replied:
Added.

majastrz reviewed (changes requested):
I added some comments on paging and examples. Please take a look. (Also, a lot of the automated checks are failing.)

KranthiKiranerusu committed:
Moved to different files

AutorestCI commented:
/azp run automation - sdk

KranthiKiranerusu committed:
Fixing Model validation

AutorestCI commented:
/azp run automation - sdk

KranthiKiranerusu commented:
@majastrz , I fixed all the auto checks except for prettier. Let me know if you see any other issues.

KranthiKiranerusu committed:
added xmspageable

AutorestCI commented:
/azp run automation - sdk

KranthiKiranerusu committed:
Fixing example

AutorestCI commented:
/azp run automation - sdk

majastrz reviewed (approved):
Signed off from ARM side with some comments:
- please make sure that every operation has a `default` response (some do and I saw some that don't)
- the paging sets nextLink to null, which will make it harder for you to add paging in the future (you should set it to nextLink to "nextLink" instead and declare the property in your list responses (even if the service returns one page only for now)

KranthiKiranerusu committed:
Added nextlinks and corrected the versions in the examples

KranthiKiranerusu commented:
> I added some comments on paging and examples. Please take a look. (Also, a lot of the automated checks are failing.)

AutorestCI commented:
/azp run automation - sdk

KranthiKiranerusu commented:
> Signed off from ARM side with some comments:
> 
> * please make sure that every operation has a `default` response (some do and I saw some that don't)
> * the paging sets nextLink to null, which will make it harder for you to add paging in the future (you should set it to nextLink to "nextLink" instead and declare the property in your list responses (even if the service returns one page only for now)

Added nextLink.

majastrz commented:
Removing the sign off because the resource tracked resources have additional requirements.

majastrz reviewed (changes requested):
Up until the last commit, this was basically a tenant-level proxy resource. Those have some of the rules relaxed and can have a list only operation, which is why I signed off.

However, the last commit made it a tracked resource (at least in the corresponding manifest PR and updated the scope in this one. Tracked resources have a number of additional requirements above proxy only:
- They have to be created via ARM, so they require PUT to be implemented
- They have to support PATCH that allows at least patching of tags
- They have to support GET that returns a single resource by name
- They have to support a list call at subscription scope.

These rules are rather hard and violating them breaks a bunch of things in the platform and also breaks some user expectations. So I don't think I can make an exception.

I'm sensing that there's some sort of misunderstanding here going on here. @arluca and @KranthiKiranerusu, does `privateLinkForAzureAD` actually need to be a tracked resource? Could it be an RG or Subscription-scoped proxy resource? Who is supposed to create resources of `privateLinkForAzureAD` type?

The other part that will complicate the resolution of this is that I will be offline on vacation for the entirety of next week. The next on-call is @ravbhatnagar so he would have to take over the reviews. I know you mentioned that you have some time pressure here, so I will stay online this weekend if you are able to respond by then.

KranthiKiranerusu committed:
Added PUt/PATCH/DELETE to comply with a trackedResource

AutorestCI commented:
/azp run automation - sdk

majastrz reviewed (approved):
Signed off from ARM side.

leni-msft commented:
Hi @KranthiKiranerusu, you need to fix the ModelValidation & PrettierCheck failures to get merged. You can refer to https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/ci-fix.md

ArLucaID committed:
Fixes for Model Validation + AllTenants Property

Fixed the error messages thrown by model validation as well as introduced the "allTenants" property under privatelinkpolicy

ArLucaID committed:
Fixes for Prettier Check

Pushing fixes done by prettier check

AutorestCI commented:
/azp run automation - sdk

ArLucaID committed:
Merge branch 'master' into patch-2

AutorestCI commented:
/azp run automation - sdk

majastrz commented:
Still signed off from ARM side.

leni-msft commented:
/azp run automation - sdk
